src/circuit/circuit_frame.rs

```diff
@@ -3055,6 +3151,168 @@ fn apply_continuation<F: LurkField, CS: ConstraintSystem<F>>(
 
         let res = AllocatedPtr::from_parts(&res_tag, &val);
 
+        // Next constraints are used for number comparisons: <, <=, >, >=
+        ///////////////////////////////////////////////////////////////////////
+        let double_a = constraints::add(&mut cs.namespace(|| "double a"), a, a)?;
+        // Ideally we would compute the bit decomposition for a, not for 2a,
+        // since it would be possible to use it for future purposes.
+        // TODO: replace bit decompositon from 2a to a.
+        let double_a_bits = double_a
+            .to_bits_le_strict(&mut cs.namespace(|| "double a lsb"))
+            .unwrap();
+        let lsb_2a = double_a_bits.get(0);
+
+        let double_b = constraints::add(&mut cs.namespace(|| "double b"), b, b)?;
+        // TODO: replace bit decompositon from 2b to b.
+        let double_b_bits = double_b
+            .to_bits_le_strict(&mut cs.namespace(|| "double b lsb"))
+            .unwrap();
+        let lsb_2b = double_b_bits.get(0);
+
+        let diff_is_zero = alloc_is_zero(&mut cs.namespace(|| "diff is zero"), &diff)?;
+        let double_diff = constraints::add(&mut cs.namespace(|| "double diff"), &diff, &diff)?;
+        // TODO: replace bit decompositon from 2diff to diff.
+        let double_diff_bits = double_diff.to_bits_le_strict(&mut cs).unwrap();
+        let lsb_2diff = double_diff_bits.get(0);
+
+        // We have that a number is defined to be negative if the parity bit (the
+        // least significant bit) is odd after doubling, meaning that the field element
+        // (after doubling) is larger than the underlying prime p that defines the
+        // field, then a modular reduction must be carried out, changing the parity that
+        // should be even (since we multiplied by 2) to odd. In other words, we define
+        // negative numbers to be those field elements that are larger than p/2.
+        let a_is_negative = lsb_2a.unwrap();
+        let b_is_negative = lsb_2b.unwrap();
+        let diff_is_negative = lsb_2diff.unwrap();
+
+        let diff_is_negative_or_zero = constraints::or(
+            &mut cs.namespace(|| "diff is negative or zero"),
+            lsb_2diff.unwrap(),
+            &diff_is_zero,
+        )?;
+
+        let diff_is_strictly_positive = Boolean::and(
+            &mut cs.namespace(|| "diff is strictly positive"),
+            &diff_is_negative.not(),
+            &Boolean::not(&diff_is_zero),
+        )?;
+
+        let diff_is_positive_or_zero = constraints::or(
+            &mut cs.namespace(|| "diff is positive or zero"),
+            &diff_is_negative.not(),
+            &diff_is_zero,
+        )?;
+
+        let both_same_sign = Boolean::xor(
+            &mut cs.namespace(|| "both same sign"),
+            a_is_negative,
+            b_is_negative,
+        )?
+        .not();
+
+        let a_negative_and_b_positive = Boolean::and(
```

Comment:
This is a misleading name. Since we are using extremely descriptive and precise names, we should also make them accurate. Otherwise it will be very hard to verify logic that uses these values.

To be clear, the *actual* constraint here is that a is negative and b is not negative. As written, b could be zero. Simply reading the code sequentially, it's not obvious whether the constraint is wrong or the naming (variable and namespace) is. Making them agree would remove this confusion about intent.

Reply:
I pushed a commit with some tests which might have failed (they don't), if this issue directly impacted correctness. I still think it's important that we treat precise names as being exactly descriptive of what they represent, though.